### PR TITLE
Filter for User Management

### PR DESCRIPTION
**Pull request by sahsisunny**
### What is the change?

Following files are modified :

- `users/constants.js` : Added the constant for filters and search button.
- `users/index.html` : Added the search button and filters.
- `users/style.css` : Added the style for search button and filters.
- `users/script.js` : Added the script for search button and filters.

### Is Development Tested?

- [x] Yes
- [ ] No

### After Change Screenshots

https://user-images.githubusercontent.com/70854507/220018420-af39edf7-38eb-4638-98d2-787d221db619.mp4

**Note :** In this PR I have added the two filters but due to some issue Skill filter is not working (UI is added and hidden). I will fix it in the next PR after the issue is resolved.

### CODE REVIEW

**Comment by heyrandhir** [comment]
umFilter

when filter modal appears the table tile viewer viewer shifts towards right.

**Comment by sahsisunny** [comment]
> umFilter
> 
> when filter modal appears the table tile toggler shifts towards right. Also the Filter Button can be moved towards right from where modal appears


https://user-images.githubusercontent.com/70854507/220171641-9ef2b5a6-3a4e-42f3-9ded-56060c5c1f6b.mp4

**Review by heyrandhir — CHANGES_REQUESTED** [review]

**Comment by heyrandhir** [comment]
In the filter button along can we add the funnel icon instead of triangle.

File: users/script.js [hunk 1]
```diff
@@ -264,6 +273,138 @@ async function getParticularUserData(
...
+
+  skillFilter.appendChild(fragment);
+  skillFilter.options[0].disabled = true;
+  skillFilter.options[0].selected = true;
+}
+
+async function getUsersStatusData(state) {
+  const usersRequest = await makeApiCall(
+    `${RDS_API_USERS}status?state=${state}`,
+  );
```
[the remark this answers is not in the document]
Comment (sahsisunny) [reply users/script.js]: done

File: users/script.js [hunk 2]
```diff
@@ -264,6 +273,138 @@ async function getParticularUserData(
...
+    { value: 'javascript', text: 'JAVASCRIPT' },
+    { value: 'typescript', text: 'TYPESCRIPT' },
+    { value: 'react', text: 'REACT JS' },
+    { value: 'vue', text: 'VUE JS' },
+    { value: 'node', text: 'NODE JS' },
+    { value: 'ember', text: 'EMBER JS' },
+    { value: 'java', text: 'JAVA' },
+    { value: 'python', text: 'PYTHON' },
+    { value: 'django', text: 'DJANGO' },
+  ];
```
[the remark this answers is not in the document]
Comment (sahsisunny) [reply users/script.js]: removed

File: users/script.js [hunk 3]
```diff
@@ -264,6 +273,138 @@ async function getParticularUserData(
   }
 }
 
+function addAvailibilityFilterOptions() {
+  const options = [
+    { value: 'none', text: 'NONE' },
+    { value: 'active', text: 'ACTIVE' },
+    { value: 'ooo', text: 'OOO' },
+    { value: 'idle', text: 'IDLE' },
```
[the remark this answers is not in the document]
Comment (sahsisunny) [reply users/script.js]: added

File: users/script.js [hunk 4]
```diff
@@ -6,10 +6,19 @@ const userSearchElement = document.getElementById(USER_SEARCH_ELEMENT);
...
+const availabilityFilter = document.getElementById(AVAILABILITY_FILTER);
+const applyFilterButton = document.getElementById(APPLY_FILTER_BUTTON);
+const clearButton = document.getElementById(CLEAR_BUTTON);
 
 let tileViewActive = false;
 let tableViewActive = true;
 let page = 0;
+let usersStatusIDLE = [];
+let usersStatusACTIVE = [];
+let usersStatusOOO = [];
```
[the remark this answers is not in the document]
Comment (sahsisunny) [reply users/script.js]: done

File: users/index.html [hunk 5]
```diff
@@ -12,14 +12,41 @@
...
     <section id="main-section">
-      <div id="search">
-        <input id="user-search" type="text" placeholder="Search Users" />
+      <div class="search-filter">
+        <div id="search">
+          <input id="user-search" type="text" placeholder="Search Users" />
+        </div>
+        <div class="filter-button-container">
+          <button id="filter-button" class="filter-button">Filters</button>
+        </div>
```
Comment (Muralidhar22) [inline users/index.html:27]: why the button is wrapped around a div, what's the use of `filter-button-container` div

File: users/script.js [hunk 6]
```diff
@@ -264,6 +272,107 @@ async function getParticularUserData(
...
+      `${RDS_API_USERS}status?state=${state}`,
+    );
+    const currentstate = await usersRequest.json();
+    return currentstate.allUserStatus;
+  } catch (err) {
+    throw new Error(`User list request failed with error: ${err}`);
+  }
+}
+
+async function showUserList(users) {
```
Comment (Muralidhar22) [inline users/script.js:309]: why is this an async function

File: users/index.html (hunk 5 above)
Comment (sahsisunny) [reply users/index.html]: for set the position of button to right btw i removed the `filter-button-container` and set position property to button

File: users/script.js (hunk 6 above)
Comment (sahsisunny) [reply users/script.js]: sorry there is no any use of async here sorry its my mistake

Comment (Muralidhar22) [reply users/script.js]: 😅 no need of sorry.

File: users/script.js [hunk 7]
```diff
@@ -264,6 +272,107 @@ async function getParticularUserData(
...
+  const availabilityFilterValue = availabilityFilter.value;
+  if (availabilityFilterValue === 'idle') {
+    idleUsers = await getUsersStatusData(IDLE);
+    await showUserList(idleUsers);
+  } else if (availabilityFilterValue === 'active') {
+    activeUsers = await getUsersStatusData(ACTIVE);
+    await showUserList(activeUsers);
+  } else if (availabilityFilterValue === 'ooo') {
+    oooUsers = await getUsersStatusData(OOO);
+    await showUserList(oooUsers);
```
Comment (Muralidhar22) [inline users/script.js:351]: please make the change here as well, why `await`

Comment (sahsisunny) [reply users/script.js:351]: to load data then run `showUserList()` function

File: users/script.js [hunk 8]
```diff
@@ -264,6 +272,107 @@ async function getParticularUserData(
   }
 }
 
+function addAvailibilityFilterOptions() {
+  const options = [
```
Comment (shmbajaj) [inline users/script.js:276]: This can come from **constants.js**

File: users/script.js [hunk 9]
```diff
@@ -264,6 +272,107 @@ async function getParticularUserData(
...
+    paginationElement.classList.remove('pagination');
+  });
+
+  userListElement.innerHTML = '';
+  userListElement.appendChild(ulElement);
+}
+
+async function filterUserByAvailability() {
+  const availabilityFilterValue = availabilityFilter.value;
+  if (availabilityFilterValue === 'idle') {
```
Comment (shmbajaj) [inline users/script.js:343]: `Suggestion [for-next-time]:` For better readibility if number of if conditions are less.

```js
if(idle)
{
//your code here
return await someMethod();
}

if(active)
{
//your code here
return await someMethod();
}

//your code here
return await someMethod();

**Review by shmbajaj — DISMISSED** [review]

File: users/script.js (hunk 8 above)
Comment (sahsisunny) [reply users/script.js:276]: text yes

File: users/style.css [hunk 10]
```diff
@@ -165,6 +172,102 @@ li p {
...
+.clear-btn {
+  background-color: var(--white-color);
+  border: 1px solid var(--light-gray-color);
+  border-radius: 0.31rem;
+  padding: 0.31rem 0.62rem;
+  cursor: pointer;
+}
+
+.clear-btn:hover {
+  background-color: var(--red-color);
```
[the remark this answers is not in the document]
Comment (sahsisunny) [reply users/style.css:214]: It has no AO.

File: users/style.css [hunk 11]
```diff
@@ -165,6 +172,102 @@ li p {
...
+}
+
+.filter-button {
+  background-color: var(--blue-color);
+  color: var(--white-color);
+  border: none;
+  border-radius: 0.31rem;
+  cursor: pointer;
+  padding: 1rem 2rem;
+  position: absolute;
```
[the remark this answers is not in the document]
Comment (sahsisunny) [reply users/style.css:253]: When i remove the position property it adjoin to search input.

File: users/script.js (hunk 9 above)
Comment (sahsisunny) [reply users/script.js:344]: changed

**Review by shmbajaj — APPROVED** [review]

**Review by heyrandhir — APPROVED** [review]

**Pull request merged**